### PR TITLE
feat: Migrate decorators to signal queries

### PR DESCRIPTION
**Pull request by n-elhk**
Migration of the following decorators:

- @Output => output()
- @ViewChild() => viewChild()
-  @ViewChildren() => viewChildren()
-  @ContentChildren() => contentChildren()
-  @ContentChild() => contentChild()

Migrate to new methode for dependency injection :

constructor(myService:MyService){} =>  myService = inject(MyService);

### CODE REVIEW

**Comment by HarelM** [comment]
Is this PR ready for review or still in progress?
Can you change it to draft and use "request for review" when completed if this is still in progress?

**Comment by n-elhk** [comment]
Yes this PR ready for ready for review, another PR to is needed to change the @Input decorators as this requires more time

**Comment by HarelM** [comment]
Would it be better to wait for the input change as well before merging this? I would prefer to have everything in the same PR so that the code won't be partially "new style" and partially "old style".
Does it make sense?

**Comment by n-elhk** [comment]
In fact, it doesn't change much, but it's true that it's better to have everything in the same "style".

**Comment by HarelM** [comment]
Thanks for the input, I'll keep this PR open and wait for the input changes.

**Comment by n-elhk** [comment]
I made the changes , PR is ready for review

File: projects/ngx-maplibre-gl/src/lib/control/control.component.ts [hunk 1]
```diff
@@ -57,28 +58,35 @@ export class CustomControl implements IControl {
...
 })
-export class ControlComponent<T extends IControl> implements OnDestroy {
+export class ControlComponent<T extends IControl> {
+  /** Init injection */
+  private readonly mapService = inject(MapService);
+  private readonly destroyRef = inject(DestroyRef);
+
   /** Init input */
-  @Input() position?: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
+  readonly position = input<
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/control.component.ts:67]: Isn't there a type for this? Control position or something?

File: projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts [hunk 2]
```diff
@@ -1,48 +1,49 @@
...
 /**
  * `mglFullscreen` - a fullscreen control directive
- * 
+ *
  * @category Directives
  */
 @Directive({
   selector: '[mglFullscreen]',
   standalone: true,
+  host: {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts:14]: Is there a better way to do this?

File: projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts [hunk 3]
```diff
@@ -1,48 +1,49 @@
...
+        );
       });
-      this.mapService.addControl(
-        this.controlComponent.control,
-        this.controlComponent.position
-      );
     });
   }
+
+  onFullscreen() {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts:46]: Is this really needed? I think maplibre resize should work now without any hacks...?

File: projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts (hunk 2 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts:14]: I saw this in the new Angular doc:

https://angular.dev/guide/components/host-elements

image

File: projects/ngx-maplibre-gl/src/lib/control/control.component.ts (hunk 1 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/control.component.ts:67]: Yes ControlPosition, i will change it !

File: projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts [hunk 4]
```diff
@@ -1,60 +1,64 @@
...
+  readonly showCompass = input<boolean>();
+  readonly showZoom = input<boolean>();
+  readonly visualizePitch = input<boolean>();
 
-      if (this.showCompass !== undefined) {
-        options.showCompass = this.showCompass;
-      }
+  constructor() {
+    afterNextRender(() => {
+      this.mapService.mapCreated$.subscribe(() => {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts:32]: I see that this logic is duplicated between a lot of directives, can we move it to a single place somehow?

File: projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts [hunk 5]
```diff
@@ -1,69 +1,70 @@
...
+          throw new Error('Another control is already set for this control');
         }
+        const options = {
+          positionOptions: this.positionOptions(),
+          fitBoundsOptions: this.fitBoundsOptions(),
+          trackUserLocation: this.trackUserLocation(),
+          showUserLocation: this.showUserLocation(),
+        };
+
+        Object.keys(options).forEach((key: string) => {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts:53]: Other directives simply use if before each option, I think using the ifs is more readable, a bit verbose, but better, IMHO.

File: projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts [hunk 6]
```diff
@@ -1,38 +1,65 @@
...
+    { host: true }
+  );
+
   /* Init inputs */
-  @Input() maxWidth?: number;
+  readonly maxWidth = input<number>();
 
   /* Dynamic inputs */
-  @Input() unit?: 'imperial' | 'metric' | 'nautical';
+  readonly unit = input<'imperial' | 'metric' | 'nautical'>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts:37]: Can we use a type for this as well? Even typeof would be better I think.

File: projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts [hunk 7]
```diff
@@ -1,38 +1,65 @@
...
+          options.unit = unit;
+        }
+        this.controlComponent.control = new ScaleControl(options);
+        this.mapService.addControl(
+          this.controlComponent.control,
+          this.controlComponent.position()
+        );
+      });
+    });
+  }
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts:62]: What about ngonchanges? It will remain as is?

File: projects/ngx-maplibre-gl/src/lib/image/image.component.ts [hunk 8]
```diff
@@ -49,39 +53,45 @@ import { MapImageData, MapImageOptions } from '../map/map.types';
...
-  private sub: Subscription;
+  readonly imageLoaded = output<void>();
 
-  constructor(private mapService: MapService, private zone: NgZone) {}
+  private isAdded = signal(false);
+  private isAdding = signal(false);
 
   ngOnInit() {
-    this.sub = this.mapService.mapLoaded$
+    this.mapService.mapLoaded$
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/image/image.component.ts:82]: What about unsubscribing?

Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/image/image.component.ts:82]: Take until destroyed... Nevermind...

File: projects/ngx-maplibre-gl/src/lib/image/image.component.ts [hunk 9]
```diff
@@ -98,25 +108,24 @@ export class ImageComponent implements OnInit, OnDestroy, OnChanges {
   }
 
   ngOnDestroy() {
-    if (this.isAdded) {
-      this.mapService.removeImage(this.id);
-    }
-    if (this.sub) {
-      this.sub.unsubscribe();
+    if (this.isAdded()) {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/image/image.component.ts:111]: Use destroy ref?

File: projects/ngx-maplibre-gl/src/lib/layer/layer.component.ts [hunk 10]
```diff
@@ -45,142 +49,131 @@ import { EventData, LayerEvents } from '../map/map.types';
...
-  @Input() layout?: LayerSpecification['layout'];
-  /** Dynamic input */
-  @Input() paint?: LayerSpecification['paint'];
-  /** Dynamic input */
-  @Input() before?: string;
-  /** Dynamic input */
-  @Input() minzoom?: LayerSpecification['minzoom'];
-  /** Dynamic input */
-  @Input() maxzoom?: LayerSpecification['maxzoom'];
+  readonly filter = input<FilterSpecification>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/layer/layer.component.ts:76]: What about the init vs dynamic input comments? 
These are used to signal the developers which input can be changed at runtime vs init only input...

File: projects/ngx-maplibre-gl/src/lib/map/map.component.spec.ts [hunk 11]
```diff
@@ -39,62 +53,67 @@ describe('MapComponent', () => {
...
+    componentRef.setInput('style', 'style');
+    fixture.detectChanges();
   });
 
   describe('Init tests', () => {
-    it('should init', () => {
-      fixture.detectChanges();
-      expect(msSpy.setup.calls.count()).toBe(1);
-    });
+    // xit("should init", fakeAsync(() => {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/map/map.component.spec.ts:62]: Are these comments needed?

**Comment by HarelM** [comment]
I haven't finished the review yet, sorry, I will continue tomorrow...

File: projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts (hunk 2 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts:14]: Maybe we can remove it completely? See my comment below.

File: projects/ngx-maplibre-gl/src/lib/map/map.component.ts [hunk 12]
```diff
@@ -77,291 +77,243 @@ import { Subscription, firstValueFrom } from 'rxjs';
...
-  @Input() preserveDrawingBuffer?: MapOptions['preserveDrawingBuffer'];
-  /** Init input */
-  @Input() trackResize?: MapOptions['trackResize'];
-  /** Init input */
-  @Input() transformRequest?: MapOptions['transformRequest'];
-  /** Init input */
-  @Input() bounds?: MapOptions['bounds']; // Use fitBounds for dynamic input
-  /** Init input */
-  @Input() antialias?: MapOptions['antialias'];
+// Omit<MapOptions, "bearing" | "container" | "pitch" | "zoom">,
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/map/map.component.ts:80]: Is this comment relevant?

File: projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts (hunk 5 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts:53]: I think keeping the "if" option is verbose and less efficient.

What we can do is create a method with an explicit name:

`const keepAvailableObjectValues = <T extends {}>(options: T): Required<T> => {
  return Object.keys(options).reduce((acc, curr) => {
    const tkey = <keyof T>curr;
    if (options[tkey] === undefined) {
      return acc;
    }
    return { ...acc, [tkey]: options[tkey] };
  }, {} as Required<T>);
};
`

And it can be use like this: 

`
      const example = keepAvailableObjectValues<ScaleControlOptions>({
        maxWidth: this.maxWidth,
        unit: this.unit,
      });
      this.controlComponent.control = new ScaleControl(example);
`

Another advantage is that it can be reused for other directives. 

What do you think ?

File: projects/ngx-maplibre-gl/src/lib/map/map.component.ts [hunk 13]
```diff
@@ -77,291 +77,243 @@ import { Subscription, firstValueFrom } from 'rxjs';
...
-          this.idle.subscribe(() => {
-            htmlElement.setAttribute('data-idle', 'true');
-          })
-        );
-        this.subscriptions.push(
-          this.render.subscribe(() => {
-            htmlElement.removeAttribute('data-idle');
-          })
-        );
+        this.mapLoad.subscribe(() => {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/map/map.component.ts:241]: Where do we unsubscrive? Maybe add takeuntildestroyed?

File: projects/ngx-maplibre-gl/src/lib/map/map.service.spec.ts [hunk 14]
```diff
@@ -157,45 +165,42 @@ describe('MapService', () => {
...
-        done();
-      });
-    })());
+  it('should update maxPitch', (done: DoneFn) => {
+    mapEvents.mapLoad.subscribe(() => {
+      mapService.updateMaxPitch(25);
+      expect(mapService.mapInstance.getMaxPitch()).toEqual(25);
+      done();
+    });
+  });
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/map/map.service.spec.ts:203]: Can we remove the below commented out test?

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts [hunk 15]
```diff
@@ -113,36 +112,41 @@ export class PopupComponent implements OnChanges, OnDestroy, OnInit {
...
+        this.mapService.removePopupFromMarker(previousMarkerInstance);
       }
-      if (this.marker && this.marker.markerInstance && this.popupInstance) {
-        this.mapService.addPopupToMarker(
-          this.marker.markerInstance,
-          this.popupInstance
-        );
+
+      const marker = this.marker();
+      const markerInstance = marker && marker.markerInstance();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:127]:
```suggestion
      const markerInstance = marker?.markerInstance();
```

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts [hunk 16]
```diff
@@ -152,40 +156,51 @@ export class PopupComponent implements OnChanges, OnDestroy, OnInit {
...
+  getLngLat(
+    lngLat: LngLatLike | undefined,
+    feature: GeoJSON.Feature<GeoJSON.Point> | undefined
+  ) {
+    if (lngLat) {
+      return lngLat;
+    } else if (feature) {
+      return feature.geometry.coordinates as [number, number];
+    }
+    throw new Error('toto');
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:204]: toto?

File: projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts [hunk 17]
```diff
@@ -24,29 +26,29 @@ import { MapService } from '../map/map.service';
...
-  @Input() animate?: CanvasSourceSpecification['animate'];
+  readonly coordinates =
+    input.required<CanvasSourceSpecification['coordinates']>();
+  readonly canvas = input.required<CanvasSourceSpecification['canvas']>();
+  readonly animate = input<CanvasSourceSpecification['animate']>();
 
-  type: CanvasSourceSpecification['type'] = 'canvas';
-  private sourceAdded = false;
+  readonly type: CanvasSourceSpecification['type'] = 'canvas';
+  private readonly sourceAdded = signal(false);
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts:44]: Why use a signal here? This looks like a private property...

File: projects/ngx-maplibre-gl/src/lib/source/geojson/feature.component.ts [hunk 18]
```diff
@@ -1,53 +1,53 @@
...
-  @Input() geometry: GeoJSON.GeometryObject;
-  /** Init input */
-  @Input() properties: any;
-  type: 'Feature' = 'Feature';
+  readonly id = model<number>(); // FIXME number only for now https://github.com/mapbox/mapbox-gl-js/issues/2716
+  readonly geometry = input.required<GeoJSON.GeometryObject>();
+  readonly properties = input<any>();
 
-  private feature: GeoJSON.Feature<GeoJSON.GeometryObject>;
+  readonly type: 'Feature' = 'Feature';
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/geojson/feature.component.ts:36]: I'm not sure this is interesting in terms of keeping as a property, I would hard code it in the relevant places.

File: projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts [hunk 19]
```diff
@@ -38,71 +40,60 @@ import { MapService } from '../../map/map.service';
...
+  readonly clusterProperties =
+    input<GeoJSONSourceSpecification['clusterProperties']>();
+  readonly lineMetrics = input<GeoJSONSourceSpecification['lineMetrics']>();
+  readonly generateId = input<GeoJSONSourceSpecification['generateId']>();
+  readonly promoteId = input<GeoJSONSourceSpecification['promoteId']>();
+  readonly filter = input<GeoJSONSourceSpecification['filter']>();
 
   /** @hidden */
-  type: GeoJSONSourceSpecification['type'] = 'geojson';
+  readonly type: GeoJSONSourceSpecification['type'] = 'geojson';
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts:84]: Same here...

File: projects/showcase/src/app/demo/examples/cluster-html.component.ts [hunk 20]
```diff
@@ -36,7 +36,7 @@ import {
   imports: [NgStyle],
 })
 export class ClusterPointComponent implements OnInit {
-  @Input() properties: any;
+  properties = input<any>();
```
Comment (HarelM) [inline projects/showcase/src/app/demo/examples/cluster-html.component.ts:39]: Can we improve the typings here?

File: projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts (hunk 5 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts:53]: This sound like a good solution, we can have this method in a util file and use it in multiple places.

File: projects/ngx-maplibre-gl/src/lib/map/map.component.ts (hunk 13 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/map/map.component.ts:241]: It's not an observable, so we need to use outputToObservable() to convert into an observable and use takeuntilDestroy. But this is not necessary, Angular destroys it itself: 

image

**Comment by HarelM** [comment]
I've finished my review.
Thanks for all the work around this PR!!
A few general comments:
I see places that use `destroyRef.onDestory` and other places which we use ngOnDestory, I think we should use the same for everything if possible.

There are places that uses signal for what were private properties, can you elaborate on my use them in this case?

This is a very large change, I've changed in my app a places with takeuntildestoried vs subsciption list and got a nasty error.
How do you know everything is working as expected? I've not sure the tests coverage here is really covering all the components unfortunately...

File: projects/ngx-maplibre-gl/src/lib/map/map.component.ts (hunk 13 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/map/map.component.ts:241]: Ahh, great, thanks!

File: projects/ngx-maplibre-gl/src/lib/map/map.service.ts [hunk 21]
```diff
@@ -420,39 +398,29 @@ export class MapService {
...
-      markerInstance.on('dragstart', (event: { target: Marker }) => {
-        if (event) {
-          const { target } = event;
-          this.zone.run(() => {
-            marker.markersEvents.markerDragStart.emit(target);
-          });
-        }
-      });
-    }
+    markerInstance.on('dragstart', (event: { target: Marker }) => {
```
[the remark this answers is not in the document]
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/map/map.service.ts:401]: "observers" does not exist in an OutputEmitterRef

[the remark this answers is not in the document]
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/map/map.service.ts:401]: Is there an equivalent or this is fine?

[the remark this answers is not in the document]
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/map/map.service.ts:405]: no equivalent, I believe that the previous "if" on observers was not necessary and the current version is correct

File: projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts (hunk 3 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/fullscreen-control.directive.ts:46]: honestly I don't know, I prefer to keep it like this

File: projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts (hunk 4 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts:33]: The logic is duplicated in some directives but I think moving it to one place makes the code less readable and more complex

File: projects/ngx-maplibre-gl/src/lib/image/image.component.ts (hunk 9 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/image/image.component.ts:113]: DestroyRef is used in the constructor

File: projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts (hunk 7 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/control/scale-control.directive.ts:60]: Some onChanges can be replaced by signal effects but I think it would be better to do that in a PR because it would require even more modifications

**Comment by n-elhk** [comment]
I saw that the unit and integration tests did not cover the entire app to check if I didn't break anything. I compared it with what is in production by replaying the showcase.

File: projects/ngx-maplibre-gl/src/lib/image/image.component.ts (hunk 9 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/image/image.component.ts:113]: I mean instead of unsubscribe...

File: projects/ngx-maplibre-gl/src/lib/layer/layer.component.ts (hunk 10 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/layer/layer.component.ts:75]: This was not addressed...
The comments are relevant to the documentation generation...

**Comment by HarelM** [comment]
There are a few last comments. Sorry for the long review...

File: projects/ngx-maplibre-gl/src/lib/image/image.component.ts (hunk 9 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/image/image.component.ts:116]: There is no more unsubscribe

**Comment by HarelM** [comment]
Can we just composition instead of inheritance for the sources?

**Comment by n-elhk** [comment]
Done

**Comment by HarelM** [comment]
Ok, I'll do another final review, if I won't see anything critical I'll merge this. 
Thanks for all the work on this @n-elhk !!

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts [hunk 22]
```diff
@@ -41,67 +43,70 @@ import { MarkerComponent } from '../marker/marker.component';
...
-  @Input() className?: PopupOptions['className'];
+export class PopupComponent implements OnChanges, OnInit {
+  /** Init injection */
+  private readonly destroyRef = inject(DestroyRef);
+  private readonly mapService = inject(MapService);
+
   /** Init input */
-  @Input() maxWidth?: PopupOptions['maxWidth'];
+  readonly closeButton = input<PopupOptions['closeButton']>();
+  readonly closeOnClick = input<PopupOptions['closeOnClick']>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:53]: There are missing comments here that were before about init input and dynamic input.
Can you go over the files and add the init/dynamic input comment for every input parameter please?

File: projects/ngx-maplibre-gl/src/lib/shared/index.ts [hunk 23]
```diff
@@ -0,0 +1 @@
+export * from './utils';
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/shared/index.ts:1]: Is this file really needed?

File: projects/ngx-maplibre-gl/src/lib/shared/utils/index.ts [hunk 24]
```diff
@@ -0,0 +1 @@
+export * from './functions';
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/shared/utils/index.ts:1]: Is this file really needed?

File: projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts [hunk 25]
```diff
@@ -1,95 +1,78 @@
...
-          this.init();
-        });
-      this.sub.add(sub);
-    });
-    this.sub.add(sub1);
+  constructor() {
+    this.sourceDirective.loadSource$
+      .pipe(
+        tap(() =>
+          this.sourceDirective.addSource(this.getCanvasSourceSpecification() as any)
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts:45]: Why do we need this "as any"?

File: projects/ngx-maplibre-gl/src/lib/source/geojson/feature.component.ts [hunk 26]
```diff
@@ -1,53 +1,55 @@
...
-export class FeatureComponent
-  implements OnInit, OnDestroy, GeoJSON.Feature<GeoJSON.GeometryObject> {
+export class FeatureComponent implements OnInit, OnDestroy {
+  /** Init injection */
+  private readonly geoJSONSourceComponent = inject<GeoJSONSourceComponent>(
+    forwardRef(() => GeoJSONSourceComponent)
+  );
   /** Init input */
-  @Input() id?: number; // FIXME number only for now https://github.com/mapbox/mapbox-gl-js/issues/2716
+  readonly id = model<number>(); // FIXME number only for now https://github.com/mapbox/mapbox-gl-js/issues/2716
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/geojson/feature.component.ts:31]: I think this comment should be removed, I don't think someone will fix this in the near future...

File: projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts [hunk 27]
```diff
@@ -38,81 +39,89 @@ import { MapService } from '../../map/map.service';
...
+        ),
+        switchMap(() => this.updateFeature()),
+        takeUntilDestroyed()
+      )
+      .subscribe();
   }
 
   ngOnChanges(changes: SimpleChanges) {
-    if (!this.sourceAdded) {
+    if (!this.sourceDirective.sourceId()) {
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts:124]: Are we sure about the removal of "sourceAdded" variable?

Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts:124]: This was moved to the directive, I see, maybe a different name like `SourceDirectiveHelper` can convey what it does better?

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+  signal,
+} from '@angular/core';
+import { Subject, fromEvent } from 'rxjs';
+import { filter, switchMap, tap } from 'rxjs/operators';
+import { MapService } from '../map/map.service';
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+import { Source, SourceSpecification } from 'maplibre-gl';
+
+/**
+ * `mgl-canvas-source` - a canvas source component
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:16]: Probably not the right comment...?

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+
+  /**  Init input */
+  readonly id = input.required<string>();
+  readonly sourceId = signal<string | null>(null);
+
+  private readonly loadSourceSubject = new Subject<void>();
+  readonly loadSource$ = this.loadSourceSubject.asObservable();
+
+  constructor() {
+    this.destroyRef.onDestroy(() => this.removeSource.bind(this));
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: Do we really need `bind(this)`? Can't we simply use `() => this.removeSource()`

Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: Do we really need `bind(this)`? Can't we simply use `() => this.removeSource()`

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+import { Source, SourceSpecification } from 'maplibre-gl';
+
+/**
+ * `mgl-canvas-source` - a canvas source component
+ * @see [canvas](https://maplibre.org/maplibre-style-spec/sources/#canvas)
+ *
+ * @category Source Components
+ */
+@Directive({
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:21]: Does this helper has to be a directive?

File: projects/showcase/src/app/demo/examples/live-update-image-srource.component.ts [hunk 31]
```diff
@@ -39,19 +45,21 @@ import data from './hike.geo.json';
...
   coordinates: number[][];
 
-  constructor(private ngZone: NgZone) {
+  constructor() {
     this.center = data.features[0].geometry!.coordinates[0];
     this.coordinates = this.makeRectangle(this.center);
     afterNextRender(() => {
       const points = data.features[0].geometry!.coordinates;
-      const coordinates = points.map((c) => this.makeRectangle(c));
+      const coordinates = points.map((c: any) => this.makeRectangle(c));
```
Comment (HarelM) [inline projects/showcase/src/app/demo/examples/live-update-image-srource.component.ts:62]:
```suggestion
      const coordinates = points.map((c: number[]) => this.makeRectangle(c));
```

**Comment by HarelM** [comment]
Can you explain the change from a local property in a class to a signal type? What's the pros and cons of this approach?

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts (hunk 29 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: yes, without "bind" its make an error in mapService

File: projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts (hunk 25 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/source/canvas-source.component.ts:45]: CanvasSourceSpecification is not part of SourceSpecification

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts (hunk 30 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:21]: Yes, its used as hostDirective in other component

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts (hunk 29 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: Did you try it with the lambda?

Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: yes it doesn't work

File: projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts (hunk 27 above)
Comment (n-elhk) [reply projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts:124]: I'm not very convinced to name it as is for two reasons: 

The first one: it is not really a helper
The second: Angular convention means that a directive must end with "Directive" like ExampleDirective.

File: projects/ngx-maplibre-gl/src/lib/source/source.directive.ts (hunk 29 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/source/source.directive.ts:37]: Can you think of a reason why this would create this error? It's very strange to me...

File: projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts (hunk 27 above)
Comment (HarelM) [reply projects/ngx-maplibre-gl/src/lib/source/geojson/geojson-source.component.ts:124]: Fair enough.

File: projects/ngx-maplibre-gl/src/lib/control/attribution-control.directive.ts [hunk 32]
```diff
@@ -1,51 +1,53 @@
...
+  /* Init injection */
+  private readonly mapService = inject(MapService);
+  private readonly controlComponent = inject<
+    ControlComponent<AttributionControl>
+  >(ControlComponent, { host: true });
+
   /** Init input */
-  @Input() customAttribution?: string | string[];
+  readonly compact = input<boolean>();
+  readonly customAttribution = input<string | string[]>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/attribution-control.directive.ts:31]: missing `/** Init input */` above this, I think.

File: projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts [hunk 33]
```diff
@@ -1,69 +1,70 @@
...
+    ControlComponent<GeolocateControl>
+  >(ControlComponent, { host: true });
+
   /* Init inputs */
-  @Input() positionOptions?: PositionOptions;
-  @Input() fitBoundsOptions?: FitBoundsOptions;
-  @Input() trackUserLocation?: boolean;
-  @Input() showUserLocation?: boolean;
+  readonly positionOptions = input<PositionOptions>();
+  readonly fitBoundsOptions = input<FitBoundsOptions>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/geolocate-control.directive.ts:39]: Missing `/** Init input */` here too, although it was missing before this change as well...

File: projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts [hunk 34]
```diff
@@ -1,60 +1,50 @@
...
-      }
-
-      this.controlComponent.control = new NavigationControl(options);
-      this.mapService.addControl(
-        this.controlComponent.control,
-        this.controlComponent.position
-      );
+  /* Init inputs */
+  readonly showCompass = input<boolean>();
+  readonly showZoom = input<boolean>();
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/control/navigation-control.directive.ts:28]: Missing `/** Init input */`, although it was missing before as well.

**Comment by HarelM** [comment]
I think this looks good.
Thanks for all the hard work on this!
Is there anything else that should be done before we merge this?

**Comment by marcjulian** [comment]
This looks awesome. 

I tested it with zoneless `provideExperimentalZonelessChangeDetection()` and the map works after removing `this.zone.onStable.pipe(first()).subscribe(() => {` in `MapService.setup`.

**Comment by n-elhk** [comment]
For me, everything seems good. As to suggest @marcjulian we can delete zone.onstable so that we can use the lib in zoneless

**Comment by marcjulian** [comment]
Yes I guess we can remove zone.onstable in another PR 

👍

**Comment by HarelM** [comment]
Last two comments:
1. Please add a changelog entry
2. If you can answer the last open review regarding the on destroy. https://github.com/maplibre/ngx-maplibre-gl/pull/178#discussion_r1661241243

THANKS!

**Comment by HarelM** [comment]
It would also be good to upgrade angular to 18, but again, as part of a different PR.

**Comment by marcjulian** [comment]
> It would also be good to upgrade angular to 18, but again, as part of a different PR.

Its already updated to [Angular 18](https://github.com/maplibre/ngx-maplibre-gl/pull/171)

**Comment by HarelM** [comment]
Right... Then we need to release a new version to match it, maybe even before this PR merge to be able to test it separately...

**Comment by HarelM** [comment]
I've released the 18.x version. Merging this now.
